src/CommunityToolkit.Maui/Views/GravatarImageSource.shared.cs

```diff
@@ -0,0 +1,238 @@
+	public override string ToString()
+	{
+		return $"Uri: {Uri}, Email:{Email}, Size:{GravatarSize}, Image: {DefaultGravatarName(Image)}, CacheValidity:{CacheValidity}, CachingEnabled:{CachingEnabled}";
+	}
+
+	/// <summary>On parent set.</summary>
+	protected override void OnParentSet()
+	{
+		object element = Parent;
+		base.OnParentSet();
+		if (element is Microsoft.Maui.Controls.Image parentImage)
+		{
+			var height = parentImage.Height >= 0 ? parentImage.Height : Math.Max(parentImage.HeightRequest, GravatarImageSourceDefaults.DefaultSize);
+			var width = parentImage.Width >= 0 ? parentImage.Width : Math.Max(parentImage.WidthRequest, GravatarImageSourceDefaults.DefaultSize);
+			var sizeFromParent = (int)Math.Max(width, height);
+			if (GravatarSize != sizeFromParent)
+			{
+				GravatarSize = sizeFromParent;
+				HandleSizeChanged(sizeFromParent);
+			}
+		}
+	}
+
+	static string DefaultGravatarName(DefaultImage defaultGravatar)
+			=> defaultGravatar switch
+			{
+				DefaultImage.FileNotFound => "404",
+				DefaultImage.MysteryPerson => "mp",
+				_ => $"{defaultGravatar}".ToLower(),
+			};
+
+	static string GetMd5Hash(string? str)
+	{
+		if (string.IsNullOrWhiteSpace(str))
+		{
+			str = "MCT";
+		}
+
+		using var md5 = MD5.Create();
+		Span<byte> hash = md5.ComputeHash(Encoding.UTF8.GetBytes(str));
+		return BitConverter.ToString(hash.ToArray(), 0, hash.Length).Replace("-", string.Empty, StringComparison.OrdinalIgnoreCase).ToLowerInvariant();
+	}
+
+	static void OnDefaultImagePropertyChanged(BindableObject bindable, object oldValue, object newValue)
+	{
+		GravatarImageSource gravatarImageSource = (GravatarImageSource)bindable;
+		gravatarImageSource.HandleDefaultImageChanged((DefaultImage)newValue);
+	}
+
+	static void OnEmailPropertyChanged(BindableObject bindable, object oldValue, object newValue)
+	{
+		GravatarImageSource gravatarImageSource = (GravatarImageSource)bindable;
+		gravatarImageSource.HandleEmailChanged((string?)newValue);
+	}
+
+	async Task<Stream> DownloadStreamAsync(Uri uri, CancellationToken cancellationToken)
+	{
+		try
+		{
+			using var client = new HttpClient();
+			// Do not remove this await otherwise the client will dispose before the stream even starts
+			return await StreamWrapper.GetStreamAsync(uri, cancellationToken, client).ConfigureAwait(false);
+		}
+		catch (Exception ex)
+		{
+			Application.Current?.FindMauiContext()?.CreateLogger<GravatarImageSource>()?.LogWarning(ex, "Error getting stream for {Uri}", Uri);
+#nullable disable
+			return null;
+#nullable enable
+		}
+	}
+
+	async Task<Stream> GetStreamAsync(Uri uri, CancellationToken cancellationToken = default)
+	{
+		cancellationToken.ThrowIfCancellationRequested();
+		return await DownloadStreamAsync(uri, cancellationToken).ConfigureAwait(false);
+	}
+
+	void HandleDefaultImageChanged(DefaultImage newValue)
+	{
+		Uri = new Uri($"https://www.gravatar.com/avatar/{GetMd5Hash(Email)}?s={GravatarSize}&d={DefaultGravatarName(newValue)}");
```

Comment:
Hmm, if it's just a URI, can we use UriImageSource?

Reply:
@VladislavAntonyuk Ideally and was my original hope, but `UriImageSource` is sealed. I had a chat with a .NET MAUI developer, and they said to inherit from ImageSource and implement an ImageSourceService.